jib-core/src/main/java/com/google/cloud/tools/jib/ncache/DefaultCacheStorageWriter.java

```diff
@@ -0,0 +1,106 @@
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.ncache;
+
+import com.google.cloud.tools.jib.blob.BlobDescriptor;
+import com.google.cloud.tools.jib.blob.Blobs;
+import com.google.cloud.tools.jib.filesystem.FileOperations;
+import com.google.cloud.tools.jib.hash.CountingDigestOutputStream;
+import com.google.cloud.tools.jib.image.DescriptorDigest;
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.zip.GZIPOutputStream;
+
+/** Writes to the default cache storage engine. */
+class DefaultCacheStorageWriter {
+
+  private final DefaultCacheStorageFiles defaultCacheStorageFiles;
+
+  DefaultCacheStorageWriter(DefaultCacheStorageFiles defaultCacheStorageFiles) {
+    this.defaultCacheStorageFiles = defaultCacheStorageFiles;
+  }
+
+  /**
+   * Writes the {@link CacheWrite}.
+   *
+   * @param cacheWrite the {@link CacheWrite} to write out
+   * @return the {@link CacheEntry} representing the written entry
+   * @throws IOException if an I/O exception occurs
+   */
+  CacheEntry write(CacheWrite cacheWrite) throws IOException {
+    Path temporaryLayerFile = defaultCacheStorageFiles.createTemporaryFile();
+
+    // Writes the UnwrittenLayer layer BLOB to a file to convert into a CachedLayer.
+    try (CountingDigestOutputStream compressedDigestOutputStream =
+        new CountingDigestOutputStream(
+            new BufferedOutputStream(FileOperations.newLockingOutputStream(temporaryLayerFile)))) {
+      // Writes the layer with GZIP compression. The original bytes are captured as the layer's
+      // diff ID and the bytes outputted from the GZIP compression are captured as the layer's
+      // content descriptor.
+      GZIPOutputStream compressorStream = new GZIPOutputStream(compressedDigestOutputStream);
+      DescriptorDigest layerDiffId =
+          cacheWrite.getLayerBlob().writeTo(compressorStream).getDigest();
+
+      // The GZIPOutputStream must be closed in order to write out the remaining compressed data.
+      compressorStream.close();
+      BlobDescriptor compressedBlobDescriptor = compressedDigestOutputStream.toBlobDescriptor();
+      DescriptorDigest layerDigest = compressedBlobDescriptor.getDigest();
+
+      // Renames the temporary layer file to the correct filename.
+      Path layerFile = defaultCacheStorageFiles.getLayerFile(layerDigest, layerDiffId);
+      Files.createDirectories(layerFile.getParent());
+      try {
+        Files.move(temporaryLayerFile, layerFile);
```

Comment:
Could moving this file while we have an output stream open on it cause problems? Can we close the output stream earlier?